types/validator_set.go

```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+				existing = existing[1:]
+			}
+			updates = updates[1:]
+		}
+		i++
+	}
+	for j := 0; j < len(existing); j++ {
+		merged[i] = existing[j]
+		i++
+	}
+	for j := 0; j < len(updates); j++ {
+		merged[i] = updates[j]
+		i++
+	}
+
+	vals.Validators = merged[:i]
+	vals.totalVotingPower = 0
+}
+
+// Checks that the validators to be removed are part of the validator set.
+func verifyRemovals(deletes []*Validator, vals *ValidatorSet) error {
+
+	for _, valUpdate := range deletes {
+		address := valUpdate.Address
+		_, val := vals.GetByAddress(address)
+		if val == nil {
+			return fmt.Errorf("failed to find validator %X to remove", address)
+		}
+	}
+	return nil
+}
+
+// Removes the validators specified in 'deletes' from validator set 'vals'.
+// Should not fail as verification has been done before.
+func (vals *ValidatorSet)applyRemovals(deletes []*Validator) {
+
+	for _, valUpdate := range deletes {
+		address := valUpdate.Address
+		_, removed := vals.Remove(address)
+		if !removed {
+			// Should never happen
+			panic(fmt.Sprintf("failed to remove validator %X", address))
+		}
+	}
+}
+
+// UpdateWithChangeSet attempts to update the validator set with 'changes'
+// It performs the following steps:
+// - validates the changes making sure there are no duplicates and
+// separates them in updates and deletes
+// - verifies that applying the updates will not result in errors (e.g. overflows)
+// - computes the priorities of validators that are added and/or changed against the final set
+// - verifies that applying the removals will not result in errors
+//   Note: currently an error is issued if a validator to be removed is not present in the set
+// - applies the updates against the validator set and performs scaling of priority values
+// - applies the removals against the validator set and performs scaling and centering of priority values
+// If error is detected during verification steps an error is returned and the validator set
+// is not changed.
+func (vals *ValidatorSet) UpdateWithChangeSet(changes []*Validator) error {
+
+	if len(changes) <= 0 {
+		return nil
+	}
+
+	// Check for duplicates within changes, split in 'updates' and 'deletes' lists (sorted)
+	updates, deletes, err := processChanges(changes)
+	if err != nil {
+		return err
+	}
+
+	// Verify that applying the 'updates' against 'vals' will not result in error.
+	// If no errors, priorities of validators in 'updates' are computed.
+	if err = verifyUpdatesAndComputeNewPriorities(updates, vals); err != nil {
+		return err
+	}
+
+	// Verify that applying the 'deletes' against 'vals' will not result in error.
+	if err = verifyRemovals(deletes, vals); err != nil {
+		return err
+	}
+
+	// Apply updates and rescale
+	vals.applyUpdates(updates)
+	vals.RescalePriorities(K * vals.TotalVotingPower())
+
+	// Apply removals, rescale and recenter
+	vals.applyRemovals(deletes)
+	vals.RescalePriorities(K * vals.TotalVotingPower())
+	vals.shiftByAvgProposerPriority()
+
```

Comment:
Question: If I understand correctly, the methods `ValidatorSet.Add`, `ValidatorSet.Update` and  `ValidatorSet.Remove` should not be used anywhere, or, only in tests. We probably still want to expose them? Or should we make them unexported?

Reply:
I removed them and changed the tests. Also I learnt today from @melekes about reflect.DeepEqual() so I changed the tests to use that.

Reply:
cool! I think, `assert.Equal` uses `reflect.DeepEqual()` internally, too.